File: sdks/go/minigun.go
// Package minigun is the Go SDK for MiniGun
// (https://github.com/ranaroussi/minigun).
//
// Stdlib-only — no external imports — so it slots into any module
// without bloating its dependency graph. Targets Go 1.21+.
//
//	mg, err := minigun.New(os.Getenv("MINIGUN_API_URL"),
//	    minigun.WithToken(os.Getenv("MINIGUN_API_TOKEN")))
//	if err != nil { log.Fatal(err) }
//
//	if _, err := mg.AddContact(ctx, "newsletter", "alice@example.com",
//	    map[string]any{"first_name": "Alice"}); err != nil {
//	    log.Fatal(err)
//	}
//
//	res, err := mg.SendBulk(ctx, minigun.SendBulkArgs{
//	    List: "newsletter", Subject: "Hi", From: "Ran <r@x.com>",
//	    MD: "Hello {{first_name | 'there'}}!",
//	})
//
// Errors fall into two buckets:
//   - *TransportError — the request never completed (DNS, TLS, timeout, ctx)
//   - *APIError       — server returned non-2xx (Status + Body populated)
//
// Both implement `error` and are unwrappable; use errors.As to branch.
package minigun

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

// Unsub-mode constants for SendBulkArgs.UnsubMode.
const (
	UnsubLocal    = "local"
	UnsubRedirect = "redirect"
	UnsubExternal = "external"
)

// TransportError wraps a network / context error that prevented us
// from getting any HTTP response back.
type TransportError struct{ Err error }

func (e *TransportError) Error() string { return "minigun: transport error: " + e.Err.Error() }
func (e *TransportError) Unwrap() error { return e.Err }

// APIError is returned when the server replied with a 4xx or 5xx.
// Body is the raw decoded payload (typically map[string]any with an
// "error" key, but kept as `any` so callers can handle non-JSON
// responses without a second type-assertion).
type APIError struct {
	Status  int
	Body    any
	Message string
}

func (e *APIError) Error() string { return fmt.Sprintf("minigun: API %d: %s", e.Status, e.Message) }

// Client is one connection-pool's worth of state. Safe for concurrent
// use — the underlying *http.Client is, and we never mutate fields
// after construction.
type Client struct {
	baseURL    string
	token      string
	httpClient *http.Client
	userAgent  string
}

// Option configures a Client at construction time.
type Option func(*Client)

// WithToken sets the bearer token sent on every request. Required
// when the server has MINIGUN_API_TOKEN configured.
func WithToken(t string) Option { return func(c *Client) { c.token = t } }

// WithHTTPClient overrides the default *http.Client (10s connect /
// 120s total). Useful for injecting custom transports, proxies, or
// tracing.
func WithHTTPClient(h *http.Client) Option { return func(c *Client) { c.httpClient = h } }

// WithTimeout sets the overall request timeout on the default
// http.Client. Ignored when WithHTTPClient is also passed (in which
// case you control the timeout on your injected client).
func WithTimeout(d time.Duration) Option {
	return func(c *Client) {
		if c.httpClient == nil {
			c.httpClient = &http.Client{Timeout: d}
		} else {
			c.httpClient.Timeout = d
		}
	}
}

// WithUserAgent overrides the default UA string. Helpful when you
// want server-side logs to attribute traffic to a specific service.
func WithUserAgent(ua string) Option { return func(c *Client) { c.userAgent = ua } }

// New constructs a Client. baseURL is the API origin (no trailing
// slash required); options are applied in order.
func New(baseURL string, opts ...Option) (*Client, error) {
	if baseURL == "" {
		return nil, errors.New("minigun: baseURL is required")
	}
	c := &Client{
		baseURL:   strings.TrimRight(baseURL, "/"),
		userAgent: "minigun-go/0.1",
	}
	for _, opt := range opts {
		opt(c)
	}
	if c.httpClient == nil {
		c.httpClient = &http.Client{Timeout: 120 * time.Second}
	}
	return c, nil
}

// ---------------------------------------------------------------------
// Contacts
// ---------------------------------------------------------------------

// AddContact upserts a contact + (re-)subscribes them to a list. Safe
// to call repeatedly with the same email: existing params are merged
// and any prior unsubscribe is cleared.
func (c *Client) AddContact(ctx context.Context, list, email string, params map[string]any) (map[string]any, error) {
	body := map[string]any{"email": email}
	if params != nil {
		body["params"] = params
	} else {
		body["params"] = nil
	}
	return c.post(ctx, "/lists/"+enc(list)+"/contacts", body)
}

// UnsubscribeContact marks a contact as unsubscribed on the given
// list. Preserves the row with subscribed=0 so future re-imports
// don't silently re-subscribe. For hard-bounce / spam-complaint
// cleanup, prefer DeleteContact.
func (c *Client) UnsubscribeContact(ctx context.Context, list, email string) (map[string]any, error) {
	return c.post(ctx, "/lists/"+enc(list)+"/unsubscribe", map[string]any{"email": email})
}

// DeleteContact permanently purges a contact + every row that
// references them (subscriptions on every list + unsubscribe-event
// audit log). Use for hard-bounce / scripted cleanup; the Mailgun
// webhook does this automatically for inbound bounce/complaint
// events. Accepts either the contact id (c_XXXX...) or the email.
func (c *Client) DeleteContact(ctx context.Context, idOrEmail string) (map[string]any, error) {
	return c.do(ctx, http.MethodDelete, "/contacts/"+enc(idOrEmail), nil)
}

// ListContacts returns a page of subscribers on a list. Pass back
// cursor from the previous response to walk forward; an empty cursor
// returns the first page.
func (c *Client) ListContacts(ctx context.Context, list, cursor string, limit int) (map[string]any, error) {
	if limit <= 0 {
		limit = 50
	}
	q := url.Values{}
	q.Set("limit", fmt.Sprintf("%d", limit))
	if cursor != "" {
		q.Set("cursor", cursor)
	}
	return c.get(ctx, "/lists/"+enc(list)+"/contacts?"+q.Encode())
}

// ---------------------------------------------------------------------
// Sends
// ---------------------------------------------------------------------

// SendSingleArgs is the input to SendSingle. To/From/Subject/Company
// and one of MD/MDFile or HTML/HTMLFile are required. Company is the
// id or slug — the sending domain is resolved from it; pass Domain to
// override for this one send.
type SendSingleArgs struct {
	To           string
	From         string
	Subject      string
	Company      string
	MD           string
	MDFile       string
	HTML         string
	HTMLFile     string
	Text         string
	TextFile     string
	Template     string
	TemplateFile string
	Preheader    string
	ReplyTo      string
	Domain       string
	List         string
	TestMode     bool
}

// SendSingle sends a single transactional email. Returns immediately
// (202); the worker performs the Mailgun POST in the background —
// poll GetSend if you need the terminal status.
func (c *Client) SendSingle(ctx context.Context, args SendSingleArgs) (map[string]any, error) {
	md, err := resolveBody("MD", args.MD, args.MDFile)
	if err != nil {
		return nil, err
	}
	html, err := resolveBody("HTML", args.HTML, args.HTMLFile)
	if err != nil {
		return nil, err
	}
	text, err := resolveBody("Text", args.Text, args.TextFile)
	if err != nil {
		return nil, err
	}
	tpl, err := resolveBody("Template", args.Template, args.TemplateFile)
	if err != nil {
		return nil, err
	}
	if md == "" && html == "" {
		return nil, errors.New("minigun: either MD/MDFile or HTML/HTMLFile is required")
	}

	return c.post(ctx, "/send/single", map[string]any{
		"to":        args.To,
		"from":      args.From,
		"subject":   args.Subject,
		"preheader": args.Preheader,
		"company":   args.Company,
		"list":      args.List,
		"reply_to":  args.ReplyTo,
		"domain":    args.Domain,
		"md":        md,
		"html":      html,
		"text":      text,
		"template":  tpl,
		"test_mode": args.TestMode,
	})
}

// SendBulkArgs is the input to SendBulk. List/Subject/From and one of
// MD/MDFile or HTML/HTMLFile are required. BatchSize and ThrottleMs
// default to 500 and 1000 respectively when zero.
type SendBulkArgs struct {
	List         string
	Subject      string
	From         string
	MD           string
	MDFile       string
	HTML         string
	HTMLFile     string
	Text         string
	TextFile     string
	Template     string
	TemplateFile string
	ReplyTo      string
	Preheader    string
	Domain       string
	BatchSize    int
	ThrottleMs   int
	NotifyTo     string
	UnsubMode    string // UnsubLocal | UnsubRedirect | UnsubExternal; empty == local
	UnsubRedir   string
	UnsubURL     string
	TestMode     bool
}

// SendBulk triggers a bulk send. Returns 202 immediately with a
// send_id while the worker drives batches in the background. The
// first batch runs inline before the 202, so the response time
// scales with batch_size + Mailgun's latency, then subsequent
// batches self-chain on the server.
func (c *Client) SendBulk(ctx context.Context, args SendBulkArgs) (map[string]any, error) {
	md, err := resolveBody("MD", args.MD, args.MDFile)
	if err != nil {
		return nil, err
	}
	html, err := resolveBody("HTML", args.HTML, args.HTMLFile)
	if err != nil {
		return nil, err
	}
	text, err := resolveBody("Text", args.Text, args.TextFile)
	if err != nil {
		return nil, err
	}
	tpl, err := resolveBody("Template", args.Template, args.TemplateFile)
	if err != nil {
		return nil, err
	}
	if md == "" && html == "" {
		return nil, errors.New("minigun: either MD/MDFile or HTML/HTMLFile is required")
	}

	unsubMode := args.UnsubMode
	if unsubMode == "" {
		unsubMode = UnsubLocal
	}
	switch unsubMode {
	case UnsubLocal, UnsubRedirect, UnsubExternal:
	default:
		return nil, fmt.Errorf("minigun: UnsubMode must be %q, %q, or %q",
			UnsubLocal, UnsubRedirect, UnsubExternal)
	}
	if unsubMode == UnsubRedirect && args.UnsubRedir == "" {
		return nil, errors.New("minigun: UnsubRedir is required when UnsubMode=redirect")
	}
	if unsubMode == UnsubExternal && args.UnsubURL == "" {
		return nil, errors.New("minigun: UnsubURL is required when UnsubMode=external")
	}

	batchSize := args.BatchSize
	if batchSize == 0 {
		batchSize = 500
	}
	throttleMs := args.ThrottleMs
	if throttleMs == 0 {
		throttleMs = 1000
	}

	return c.post(ctx, "/send/bulk", map[string]any{
		"list":         args.List,
		"subject":      args.Subject,
		"from":         args.From,
		"reply_to":     args.ReplyTo,
		"preheader":    args.Preheader,
		"domain":       args.Domain,
		"md":           md,
		"html":         html,
		"text":         text,
		"template":     tpl,
		"batch_size":   batchSize,
		"throttle_ms":  throttleMs,
		"notify_email": args.NotifyTo,
		"unsub_mode":   unsubMode,
		"unsub_redir":  args.UnsubRedir,
		"unsub_url":    args.UnsubURL,
		"test_mode":    args.TestMode,
	})
}

// GetSend returns a one-shot snapshot of a send's status + per-batch
// progress.
func (c *Client) GetSend(ctx context.Context, sendID string) (map[string]any, error) {
	return c.get(ctx, "/send/"+enc(sendID))
}

// GetSendStats returns aggregate stats. DB-backed for completed
// sends; falls back to a live Mailgun Metrics API call for in-flight
// / just-finished ones.
func (c *Client) GetSendStats(ctx context.Context, sendID string) (map[string]any, error) {
	return c.get(ctx, "/send/"+enc(sendID)+"/stats")
}

// ResumeSend resumes a paused / failed send. Pass force=true ONLY
// if a batch was left in_flight — Mailgun may already have accepted
// it, so a retry can duplicate-send.
func (c *Client) ResumeSend(ctx context.Context, sendID string, force bool) (map[string]any, error) {
	path := "/send/" + enc(sendID) + "/resume"
	if force {
		path += "?force=1"
	}
	return c.post(ctx, path, map[string]any{})
}

// ListSendEventsArgs narrows what ListSendEvents returns.
// Zero values mean "no filter."
type ListSendEventsArgs struct {
	SendID  string
	Event   string // delivered | opened | clicked | failed | complained | unsubscribed
	SinceMs int64  // lower bound on event_timestamp_ms
	Limit   int    // page size (default 100, max 500)
	Cursor  string // opaque keyset cursor from a previous page's next_cursor
}

// ListSendEvents returns one page of archived Mailgun events for a
// send. Requires EVENTS_ARCHIVE_ENABLED on the server. The response
// shape is { items: [...], next_cursor?: string } — when next_cursor
// is absent, the page is the last one.
func (c *Client) ListSendEvents(ctx context.Context, a ListSendEventsArgs) (map[string]any, error) {
	if a.SendID == "" {
		return nil, errors.New("minigun: SendID is required")
	}
	q := url.Values{}
	if a.Event != "" {
		q.Set("event", a.Event)
	}
	if a.SinceMs > 0 {
		q.Set("since", fmt.Sprintf("%d", a.SinceMs))
	}
	if a.Limit > 0 {
		q.Set("limit", fmt.Sprintf("%d", a.Limit))
	}
	if a.Cursor != "" {
		q.Set("cursor", a.Cursor)
	}
	path := "/send/" + enc(a.SendID) + "/events"
	if enc := q.Encode(); enc != "" {
		path += "?" + enc
	}
	return c.get(ctx, path)
}

// GetContactEngagement returns per-list engagement counters for one
// contact. idOrEmail accepts a contact id (c_*) or email. listID,
// when non-empty, narrows to one list (accepts id or slug).
func (c *Client) GetContactEngagement(ctx context.Context, idOrEmail, listID string) (map[string]any, error) {
	if idOrEmail == "" {
		return nil, errors.New("minigun: idOrEmail is required")
	}
	path := "/contacts/" + enc(idOrEmail) + "/engagement"
	if listID != "" {
		path += "?list_id=" + url.QueryEscape(listID)
	}
	return c.get(ctx, path)
}

// PruneListArgs describes a list-hygiene run. At least one of the three
// criteria fields must be > 0 — pruning a list with no criteria would
// match every contact and is rejected server-side.
type PruneListArgs struct {
	List                       string
	MinMessagesSinceEngagement int64 // messages_since_last_engagement >= N
	DormantForDays             int64 // last open/click older than D days
	NoDeliveryForDays          int64 // never delivered in last D days
	// DryRun=nil defaults to TRUE on the wire (server-side default). Set
	// *bool=&false explicitly to commit. The two-step ergonomics are
	// intentional: callers can't accidentally purge a list by forgetting
	// to set a field.
	DryRun     *bool
	Limit      int
	SampleSize int
}

// PruneList unsubscribes dormant contacts from a list based on
// engagement signals from the events archive. Requires Phase 2
// (events archive) data on the server.
//
// DryRun=nil → server defaults to dry_run=true. Set DryRun to &false
// explicitly to commit.
//
// Returns: {list_id, dry_run, candidates, unsubscribed, sample, reason_counts}.
func (c *Client) PruneList(ctx context.Context, a PruneListArgs) (map[string]any, error) {
	if a.List == "" {
		return nil, errors.New("minigun: List is required")
	}
	if a.MinMessagesSinceEngagement <= 0 && a.DormantForDays <= 0 && a.NoDeliveryForDays <= 0 {
		return nil, errors.New("minigun: at least one of MinMessagesSinceEngagement, DormantForDays, NoDeliveryForDays must be > 0")
	}
	body := map[string]any{
		"min_messages_since_engagement": a.MinMessagesSinceEngagement,
		"dormant_for_days":              a.DormantForDays,
		"no_delivery_for_days":          a.NoDeliveryForDays,
	}
	if a.DryRun != nil {
		body["dry_run"] = *a.DryRun
	}
	if a.Limit > 0 {
		body["limit"] = a.Limit
	}
	if a.SampleSize > 0 {
		body["sample_size"] = a.SampleSize
	}
	return c.post(ctx, "/lists/"+enc(a.List)+"/prune", body)
}

// ---------------------------------------------------------------------
// Transport
// ---------------------------------------------------------------------

func (c *Client) get(ctx context.Context, path string) (map[string]any, error) {
	return c.do(ctx, http.MethodGet, path, nil)
}

func (c *Client) post(ctx context.Context, path string, body any) (map[string]any, error) {
	return c.do(ctx, http.MethodPost, path, body)
}

func (c *Client) do(ctx context.Context, method, path string, body any) (map[string]any, error) {
	var reader io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("minigun: marshal body: %w", err)
		}
		reader = bytes.NewReader(buf)
	}

	req, err := http.NewRequestWithContext(ctx, method, c.baseURL+path, reader)
	if err != nil {
		return nil, &TransportError{Err: err}
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", c.userAgent)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if c.token != "" {
		req.Header.Set("Authorization", "Bearer "+c.token)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, &TransportError{Err: err}
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &TransportError{Err: err}
	}

	var decoded any
	if len(raw) > 0 {
		// Best-effort JSON decode; non-JSON payloads (e.g. a stray
		// HTML 502 from an edge) flow through to APIError untouched.
		if jerr := json.Unmarshal(raw, &decoded); jerr != nil {
			decoded = string(raw)
		}
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		msg := ""
		if m, ok := decoded.(map[string]any); ok {
			if e, ok := m["error"].(string); ok {
				msg = e
			}
		}
		if msg == "" {
			if s, ok := decoded.(string); ok {
				msg = s
			} else {
				msg = http.StatusText(resp.StatusCode)
			}
		}
		return nil, &APIError{Status: resp.StatusCode, Body: decoded, Message: msg}
	}

	if m, ok := decoded.(map[string]any); ok {
		return m, nil
	}
	return map[string]any{}, nil
}

// ---------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------

func enc(s string) string { return url.PathEscape(s) }

// resolveBody returns whichever of `direct` or the contents of
// `file` is non-empty. Returns an error if both are non-empty, or
// the file is unreadable. Returns "" (no error) when neither is set;
// the caller decides whether that's required.
func resolveBody(name, direct, file string) (string, error) {
	if direct != "" && file != "" {
		return "", fmt.Errorf("minigun: pass only one of %s or %sFile, not both", name, name)
	}
	if file == "" {
		return direct, nil
	}
	b, err := os.ReadFile(file)
	if err != nil {
		return "", fmt.Errorf("minigun: read %sFile %q: %w", name, file, err)
	}
	return string(b), nil
}

File: sdks/python/minigun.py
"""
Python SDK for MiniGun (https://github.com/ranaroussi/minigun).

Stdlib-only — no `requests`, no `httpx` — so you can drop this single
file into any project without touching its dependency tree. Targets
Python 3.9+.

    from minigun import Minigun, MinigunApiError, MinigunTransportError

    mg = Minigun(os.environ["MINIGUN_API_URL"], os.environ["MINIGUN_API_TOKEN"])
    mg.add_contact("newsletter", "alice@example.com", {"first_name": "Alice"})
    res = mg.send_bulk(
        list="newsletter",
        subject="Hi",
        from_="Ran <r@x.com>",
        md="Hello {{first_name | 'there'}}!",
    )
    print(res["send_id"])

All methods raise:
  - MinigunTransportError on a network / SSL / timeout failure
  - MinigunApiError on any 4xx/5xx (status + decoded body are attributes)
  - ValueError on local argument validation (mutually-exclusive flags, etc.)
MinigunError is the common base, so `except MinigunError` catches both.
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Mapping, Optional

__all__ = [
    "Minigun",
    "MinigunError",
    "MinigunTransportError",
    "MinigunApiError",
    "UNSUB_LOCAL",
    "UNSUB_REDIRECT",
    "UNSUB_EXTERNAL",
]

UNSUB_LOCAL = "local"
UNSUB_REDIRECT = "redirect"
UNSUB_EXTERNAL = "external"


class MinigunError(Exception):
    """Base for every error this SDK raises after construction."""


class MinigunTransportError(MinigunError):
    """The HTTP request itself failed (DNS, TLS, timeout, refused, etc.)."""


class MinigunApiError(MinigunError):
    """The server replied with a non-2xx status."""

    def __init__(self, status: int, body: Any, message: str) -> None:
        super().__init__(message)
        self.status = status
        self.body = body


class Minigun:
    """One client per (base_url, token) pair. Thread-safe — internal
    state is read-only after __init__, and each request opens its own
    urllib opener."""

    def __init__(
        self,
        base_url: str,
        token: str = "",
        *,
        connect_timeout: float = 10.0,
        timeout: float = 120.0,
        user_agent: str = "minigun-python/0.1",
    ) -> None:
        if not base_url:
            raise ValueError("base_url is required")
        self._base_url = base_url.rstrip("/")
        self._token = token
        # urllib's `timeout=` is a single overall timeout, not separate
        # connect/read like requests. We keep both args for API parity
        # with the other SDKs but only the larger value is enforced.
        self._timeout = max(float(connect_timeout), float(timeout))
        self._user_agent = user_agent

    # -----------------------------------------------------------------
    # Contacts
    # -----------------------------------------------------------------

    def add_contact(
        self,
        list: str,
        email: str,
        params: Optional[Mapping[str, Any]] = None,
    ) -> dict:
        """Upsert a contact + (re-)subscribe them to a list.

        Safe to call repeatedly: existing rows get `params` merged and
        any prior unsubscribe is cleared.
        """
        path = f"/lists/{_q(list)}/contacts"
        return self._post(path, {"email": email, "params": dict(params) if params is not None else None})

    def unsubscribe_contact(self, list: str, email: str) -> dict:
        """Admin-side unsubscribe by email (no HMAC token required).

        Preserves the row with `subscribed=0` so future re-imports
        don't silently re-subscribe. For hard-bounce / spam-complaint
        cleanup prefer delete_contact().
        """
        path = f"/lists/{_q(list)}/unsubscribe"
        return self._post(path, {"email": email})

    def delete_contact(self, id_or_email: str) -> dict:
        """Permanently purge a contact + every row that references them
        (subscriptions on every list + unsubscribe-event audit log).

        Use for hard-bounce cleanup so the address can't be picked up
        by a future bulk send. The Mailgun webhook does this
        automatically; this method is for scripted / one-off purges.
        Accepts either the contact id (c_XXXX...) or the email address.
        """
        return self._delete(f"/contacts/{_q(id_or_email)}")

    def list_contacts(
        self,
        list: str,
        cursor: Optional[str] = None,
        limit: int = 50,
    ) -> dict:
        """Paginated subscribers on a list. Pass back `cursor` from the
        previous response to walk forward."""
        qs: dict[str, str] = {"limit": str(limit)}
        if cursor:
            qs["cursor"] = cursor
        path = f"/lists/{_q(list)}/contacts?{urllib.parse.urlencode(qs)}"
        return self._get(path)

    # -----------------------------------------------------------------
    # Sends
    # -----------------------------------------------------------------

    def send_single(
        self,
        *,
        to: str,
        from_: str,
        subject: str,
        company: str,
        md: Optional[str] = None,
        md_file: Optional[str] = None,
        html: Optional[str] = None,
        html_file: Optional[str] = None,
        text: Optional[str] = None,
        text_file: Optional[str] = None,
        template: Optional[str] = None,
        template_file: Optional[str] = None,
        preheader: Optional[str] = None,
        reply_to: Optional[str] = None,
        domain: Optional[str] = None,
        list: Optional[str] = None,
        test_mode: bool = False,
    ) -> dict:
        """Send a single transactional email.

        Required: `to`, `from_`, `subject`, `company`, and one of
        md/md_file or html/html_file. `company` is the company id or
        slug — MiniGun resolves the sending domain from it. Pass
        `domain=` to override for this one send.

        Pass at most one of each `*` / `*_file` pair. Files are read
        synchronously at call time.

        Returns immediately (202). The worker performs the Mailgun
        POST in the background — poll get_send() if you need the
        terminal status.
        """
        md = _resolve_body("md", md, md_file)
        html = _resolve_body("html", html, html_file)
        text = _resolve_body("text", text, text_file)
        template = _resolve_body("template", template, template_file)
        if md is None and html is None:
            raise ValueError("either md/md_file or html/html_file is required")

        return self._post(
            "/send/single",
            {
                "to": to,
                "from": from_,
                "subject": subject,
                "preheader": preheader or "",
                "company": company,
                "list": list or "",
                "reply_to": reply_to or "",
                "domain": domain or "",
                "md": md or "",
                "html": html or "",
                "text": text or "",
                "template": template or "",
                "test_mode": test_mode,
            },
        )

    def send_bulk(
        self,
        *,
        list: str,
        subject: str,
        from_: str,
        md: Optional[str] = None,
        md_file: Optional[str] = None,
        html: Optional[str] = None,
        html_file: Optional[str] = None,
        text: Optional[str] = None,
        text_file: Optional[str] = None,
        template: Optional[str] = None,
        template_file: Optional[str] = None,
        reply_to: Optional[str] = None,
        preheader: Optional[str] = None,
        domain: Optional[str] = None,
        batch_size: int = 500,
        throttle_ms: int = 1000,
        notify_to: Optional[str] = None,
        unsub_mode: str = UNSUB_LOCAL,
        unsub_redir: Optional[str] = None,
        unsub_url: Optional[str] = None,
        test_mode: bool = False,
    ) -> dict:
        """Trigger a bulk send to a list. Returns 202 immediately with
        a send_id while the worker drives batches in the background.

        The first batch runs inline before the 202 so the response
        time scales with batch_size + Mailgun's latency, then
        subsequent batches self-chain on the server.
        """
        md = _resolve_body("md", md, md_file)
        html = _resolve_body("html", html, html_file)
        text = _resolve_body("text", text, text_file)
        template = _resolve_body("template", template, template_file)
        if md is None and html is None:
            raise ValueError("either md/md_file or html/html_file is required")
        if unsub_mode not in (UNSUB_LOCAL, UNSUB_REDIRECT, UNSUB_EXTERNAL):
            raise ValueError("unsub_mode must be 'local', 'redirect', or 'external'")
        if unsub_mode == UNSUB_REDIRECT and not unsub_redir:
            raise ValueError("unsub_redir is required when unsub_mode='redirect'")
        if unsub_mode == UNSUB_EXTERNAL and not unsub_url:
            raise ValueError("unsub_url is required when unsub_mode='external'")

        return self._post(
            "/send/bulk",
            {
                "list": list,
                "subject": subject,
                "from": from_,
                "reply_to": reply_to or "",
                "preheader": preheader or "",
                "domain": domain or "",
                "md": md or "",
                "html": html or "",
                "text": text or "",
                "template": template or "",
                "batch_size": batch_size,
                "throttle_ms": throttle_ms,
                "notify_email": notify_to or "",
                "unsub_mode": unsub_mode,
                "unsub_redir": unsub_redir or "",
                "unsub_url": unsub_url or "",
                "test_mode": test_mode,
            },
        )

    def get_send(self, send_id: str) -> dict:
        """One-shot snapshot of a send's status + per-batch progress."""
        return self._get(f"/send/{_q(send_id)}")

    def get_send_stats(self, send_id: str) -> dict:
        """Aggregate stats. DB-backed for completed sends; falls back
        to a live Mailgun Metrics API call for in-flight or just-
        finished ones."""
        return self._get(f"/send/{_q(send_id)}/stats")

    def resume_send(self, send_id: str, force: bool = False) -> dict:
        """Resume a paused or failed send. Pass force=True ONLY if a
        batch was left in_flight: Mailgun may already have accepted
        it, so a retry can duplicate-send."""
        path = f"/send/{_q(send_id)}/resume"
        if force:
            path += "?force=1"
        return self._post(path, {})

    def list_send_recipients(
        self,
        send_id: str,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> dict:
        """One page of per-recipient message engagement for a send (one
        row per contact: sent/delivered timestamps, first/last open + click
        with counts, failure/complaint/unsubscribe state). Requires
        EVENTS_ARCHIVE_ENABLED on the server.

        Returns {"items": [...], "next_cursor"?: str}. Keyset paginated by
        contact_id.

        - limit:  page size (default 100, max 500)
        - cursor: opaque cursor from a previous page's next_cursor
        """
        params = []
        if limit:
            params.append(f"limit={limit}")
        if cursor:
            params.append(f"cursor={_q(cursor)}")
        path = f"/send/{_q(send_id)}/recipients"
        if params:
            path += "?" + "&".join(params)
        return self._get(path)

    def get_contact_engagement(
        self,
        id_or_email: str,
        list_id: Optional[str] = None,
    ) -> dict:
        """Per-list engagement counters for one contact. id_or_email
        accepts a contact id (c_*) or email. Pass list_id to narrow
        to one list (accepts id or slug)."""
        path = f"/contacts/{_q(id_or_email)}/engagement"
        if list_id:
            path += f"?list_id={_q(list_id)}"
        return self._get(path)

    def prune_list(
        self,
        list_id: str,
        min_messages_since_engagement: int = 0,
        dormant_for_days: int = 0,
        no_delivery_for_days: int = 0,
        dry_run: Optional[bool] = None,
        limit: Optional[int] = None,
        sample_size: Optional[int] = None,
    ) -> dict:
        """Unsubscribe dormant contacts from a list. dry_run defaults
        to TRUE server-side — explicitly pass dry_run=False to commit.

        At least one criterion must be > 0; multiple are OR'd.

        - min_messages_since_engagement: messages_since_last_engagement >= N
        - dormant_for_days: last open/click older than D days
        - no_delivery_for_days: never delivered to in the last D days

        Returns {list_id, dry_run, candidates, unsubscribed, sample,
        reason_counts}.
        """
        if min_messages_since_engagement <= 0 and dormant_for_days <= 0 and no_delivery_for_days <= 0:
            raise ValueError(
                "at least one of min_messages_since_engagement, "
                "dormant_for_days, no_delivery_for_days must be > 0"
            )
        body: dict = {
            "min_messages_since_engagement": min_messages_since_engagement,
            "dormant_for_days": dormant_for_days,
            "no_delivery_for_days": no_delivery_for_days,
        }
        if dry_run is not None:
            body["dry_run"] = dry_run
        if limit:
            body["limit"] = limit
        if sample_size:
            body["sample_size"] = sample_size
        return self._post(f"/lists/{_q(list_id)}/prune", body)

    # -----------------------------------------------------------------
    # Transport
    # -----------------------------------------------------------------

    def _get(self, path: str) -> dict:
        return self._request("GET", path, None)

    def _post(self, path: str, body: Any) -> dict:
        return self._request("POST", path, body)

    def _delete(self, path: str) -> dict:
        return self._request("DELETE", path, None)

    def _request(self, method: str, path: str, body: Any) -> dict:
        url = self._base_url + path
        headers = {
            "Accept": "application/json",
            "User-Agent": self._user_agent,
        }
        if self._token:
            headers["Authorization"] = "Bearer " + self._token

        data: Optional[bytes] = None
        if body is not None:
            headers["Content-Type"] = "application/json"
            data = json.dumps(body).encode("utf-8")

        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                status = resp.getcode()
                raw = resp.read()
        except urllib.error.HTTPError as e:
            # 4xx/5xx — body is still readable and we want to surface
            # it as MinigunApiError rather than the urllib transport
            # exception, which is much less useful at the callsite.
            status = e.code
            raw = e.read()
        except urllib.error.URLError as e:
            raise MinigunTransportError(f"transport error: {e.reason}") from e
        except (TimeoutError, OSError) as e:
            raise MinigunTransportError(f"transport error: {e}") from e

        decoded: Any = None
        if raw:
            try:
                decoded = json.loads(raw.decode("utf-8"))
            except (ValueError, UnicodeDecodeError):
                decoded = raw

        if not 200 <= status < 300:
            msg = (
                decoded.get("error")
                if isinstance(decoded, dict) and "error" in decoded
                else (decoded if isinstance(decoded, str) else str(decoded))
            )
            raise MinigunApiError(status, decoded, f"MiniGun API {status}: {msg}")

        return decoded if isinstance(decoded, dict) else {}


# ---------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------


def _q(s: str) -> str:
    """URL-encode a path segment. safe='' so '/' inside an email
    address (impossible but defensive) gets encoded too."""
    return urllib.parse.quote(s, safe="")


def _resolve_body(name: str, direct: Optional[str], file: Optional[str]) -> Optional[str]:
    """Resolve a body-or-file pair. Returns None when neither was
    supplied (caller decides whether the field is required); raises
    ValueError when both are supplied or the file is unreadable."""
    if direct is not None and file is not None:
        raise ValueError(f"pass only one of {name} or {name}_file, not both")
    if file is None:
        return direct
    if not os.path.isfile(file):
        raise ValueError(f"{name}_file '{file}' does not exist")
    with open(file, "r", encoding="utf-8") as fh:
        return fh.read()
